src/Microsoft.AspNetCore.Razor.Evolution/TagHelperBinderSyntaxTreePass.cs

```diff
@@ -61,5 +78,63 @@ private IReadOnlyList<RazorError> CombineErrors(IReadOnlyList<RazorError> errors
 
             return combinedErrors;
         }
+
+        private class Visitor : ParserVisitor
+        {
+            public List<TagHelperDirectiveDescriptor> Directives { get; } = new List<TagHelperDirectiveDescriptor>();
+
+            public override void VisitAddTagHelperSpan(AddTagHelperChunkGenerator chunkGenerator, Span span)
+            {
+                Directives.Add(CreateDirective(span, chunkGenerator.LookupText, TagHelperDirectiveType.AddTagHelper));
+            }
+
+            public override void VisitRemoveTagHelperSpan(RemoveTagHelperChunkGenerator chunkGenerator, Span span)
+            {
+                Directives.Add(CreateDirective(span, chunkGenerator.LookupText, TagHelperDirectiveType.RemoveTagHelper));
+            }
+
+            public override void VisitTagHelperPrefixDirectiveSpan(TagHelperPrefixDirectiveChunkGenerator chunkGenerator, Span span)
+            {
+                Directives.Add(CreateDirective(span, chunkGenerator.Prefix, TagHelperDirectiveType.TagHelperPrefix));
+            }
+
+            private TagHelperDirectiveDescriptor CreateDirective(
+                Span span,
+                string directiveText,
+                TagHelperDirectiveType directiveType)
+            {
+                directiveText = directiveText.Trim();
+
+                // If this is the "string literal" form of a directive, we'll need to postprocess the location
+                // and content.
+                //
+                // Ex: @addTagHelper "*, Microsoft.AspNetCore.CoolLibrary"
+                //                    ^                                 ^
+                //                  Start                              End
+                var directiveStart = span.Start;
+                if (span.Symbols.Count == 1 && (span.Symbols[0] as CSharpSymbol)?.Type == CSharpSymbolType.StringLiteral)
+                {
+                    var offset = span.Content.IndexOf(directiveText, StringComparison.Ordinal);
+
+                    // This is safe because inside one of these directives all of the text needs to be on the
+                    // same line.
+                    var original = span.Start;
+                    directiveStart = new SourceLocation(
+                        original.FilePath,
+                        original.AbsoluteIndex + offset,
+                        original.LineIndex,
+                        original.CharacterIndex + offset);
+                }
+
+                var directiveDescriptor = new TagHelperDirectiveDescriptor
+                {
+                    DirectiveText = directiveText,
+                    Location = directiveStart,
+                    DirectiveType = directiveType,
+                };
+
+                return directiveDescriptor;
```

Comment:
This is the only significant logic left from the original implementation. The rest was just visitor boilerplate.